### PR TITLE
Switch to `@ampproject/remapping` to merge source maps

### PR DESCRIPTION
**Pull request by jridgewell**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  | Yes
| License                  | MIT

This switches us from the custom-build source map merging to [`@ampproject/remapping`](https://github.com/ampproject/remapping/). This should result in a considerably faster merges with reduced memory usage. Eg, https://github.com/angular/angular-cli/issues/21779 found that using the old method caused OOM errors when the input/output maps were large enough.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14209"></a>

### CODE REVIEW

File: packages/babel-core/src/transformation/file/merge-map.ts [hunk 1]
```diff
@@ -1,321 +1,25 @@
...
-  const consumer = new sourceMap.SourceMapConsumer({
+function rootless(map: SourceMap): SourceMap {
+  return {
     ...map,
 
-    // This is a bit hack. .addMapping expects source values to be relative,
-    // but eachMapping returns mappings with absolute paths. To avoid that
-    // incompatibility, we leave the sourceRoot out here and add it to the
-    // final map at the end instead.
+    // This is a bit hack. Remapping will create absolute sources in our
```
Comment (nicolo-ribaudo) [inline packages/babel-core/src/transformation/file/merge-map.ts:20]: Is it true for `@ampproject/remapping` too? In the readme, the `helloworld.js` doesn't seem to be transformed to absolute.

**Comment by nicolo-ribaudo** [comment]
Could you verify if the failing tests are correct? I don't think that `names` should include identifier that aren't in the existing source map (because it means that they weren't in the original code).

File: packages/babel-core/src/transformation/file/merge-map.ts (hunk 1 above)
Comment (jridgewell) [reply packages/babel-core/src/transformation/file/merge-map.ts:20]: Unfortunately so: https://github.com/ampproject/remapping/blob/fcc9f442f43249389dc946cbe8bb3736a8033d6a/src/source-map-tree.ts#L115.

> In the readme, the helloworld.js doesn't seem to be transformed to absolute.

The readme example doesn't have a `sourceRoot`. If one were included, say `src`, then the output would partially match `{ sourceRoot: undefined, sources: ['src/helloworld.js'] }`.

**Comment by jridgewell** [comment]
I actually get 0 failures on my local, so I'm still trying to debug what's happening.

File: packages/babel-core/test/api.js [hunk 2]
```diff
@@ -540,7 +540,7 @@ describe("api", function () {
         column: 4,
       }),
     ).toEqual({
-      name: null,
+      name: "Foo",
```
Comment (jridgewell) [inline packages/babel-core/test/api.js:543]: The "Foo" name doesn't exist in the input sourcemap, but does in the transformed sourcemap. Remapping will use the input's if it existed, but falls back to the transformed one if not.

**Review by jridgewell — COMMENTED** [review]
The changed sourcemaps here all refer to the same `var foo = () => 4;` input, which only has 2 mappings:

```
var foo = () => 4;
^^^^^^^^^
         ^^^^^^^^^
```

The old code would create new mappings based on the transformed map, but these [don't actually point](http://justin.ridgewell.name/source-map-visualization/#base64,dmFyIGZvbyA9IGZ1bmN0aW9uICgpIHsKICByZXR1cm4gNDsKfTsK,eyJtYXBwaW5ncyI6IkFBQUEsSUFBQSxHQUFBLEdBQVUsWTtTQUFNLEM7QUFBQyxDQUFqQiIsIm5hbWVzIjpbXSwic291cmNlcyI6WyJvcmlnaW5hbC5qcyJdLCJzb3VyY2VzQ29udGVudCI6WyJ2YXIgZm9vID0gKCkgPT4gNDsiXSwidmVyc2lvbiI6MywiZmlsZSI6ImV4YW1wbGUuanMifQ==,dmFyIGZvbyA9ICgpID0+IDQ7) to a sane spot in the original file. See how when you hover over `foo` on the left, it hightlights `var foo = ` on the right? This is because the `foo` mapping is pointing to line 1 column 0, not column 4 (where `foo` exists in the original code). If it did, it would correctly highlight the `foo` identifier on the right side.

The new output [doesn't generate](http://justin.ridgewell.name/source-map-visualization/#base64,dmFyIGZvbyA9IGZ1bmN0aW9uICgpIHsKICByZXR1cm4gNDsKfTsK,eyJtYXBwaW5ncyI6IkFBQUEsVUFBVSxZO1NBQU0sQztBQUFDLENBQWpCIiwibmFtZXMiOltdLCJzb3VyY2VzIjpbIm9yaWdpbmFsLmpzIl0sInNvdXJjZXNDb250ZW50IjpbInZhciBmb28gPSAoKSA9PiA0OyJdLCJ2ZXJzaW9uIjozLCJmaWxlIjoiZXhhbXBsZS5qcyJ9,dmFyIGZvbyA9ICgpID0+IDQ7) a useless mapping for `foo`. But, if the input sourcemap were higher fidelity (with mappings at identifiers, syntax, etc), it would.

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

**Review by JLHwung — APPROVED** [review]

**Pull request merged**